src/psyclone/psyir/transformations/inline_trans.py

```diff
@@ -413,62 +791,139 @@ def validate(self, node, options=None):
-                        rank += 1
-                        # TODO #924 add a method to Range that returns the
-                        # PSyIR expression for the number of elements it
-                        # contains. This can then be compared with the shape of
-                        # the dummy argument.
-                        if not actual_arg.is_full_range(idx):
-                            # It's OK to use the loop variable in the lambda
-                            # definition because if we get to this point then
-                            # we're going to quit the loop.
-                            # pylint: disable=cell-var-from-loop
-                            raise TransformationError(LazyString(
-                                lambda: f"Cannot inline routine "
-                                f"'{routine.name}' because argument "
-                                f"'{visitor(actual_arg)}' is "
-                                f"an array subsection (TODO #924)."))
-                if isinstance(dummy_arg.datatype, ScalarType):
-                    dummy_rank = 0
-                else:
-                    dummy_rank = len(dummy_arg.datatype.shape)
-                if rank != dummy_rank:
-                    # It's OK to use the loop variable in the lambda definition
-                    # because if we get to this point then we're going to quit
-                    # the loop.
-                    # pylint: disable=cell-var-from-loop
-                    raise TransformationError(LazyString(
+        for formal_arg, actual_arg in zip(routine_table.argument_list,
+                                          node.children):
+            # If the formal argument is an array with non-default bounds then
+            # we also need to know the bounds of that array at the call site.
+            # For this reason, we cannot support formal arguments of
+            # UnknownType because they might be arrays with non-default bounds.
+            if isinstance(formal_arg.datatype, UnknownType):
+                raise TransformationError(
+                    f"Routine '{routine.name}' cannot be inlined because "
+                    f"formal argument '{formal_arg.name}' is of UnknownType")
+
+            if not isinstance(formal_arg.datatype, ArrayType):
+                # Formal argument is not an array so we don't need to do any
+                # further checks.
+                continue
+
+            if not isinstance(actual_arg, (Reference, Literal)):
+                # TODO #1799 this really needs the `datatype` method to be
+                # extended to support all nodes. For now we have to abort
+                # if we encounter an argument that is not a scalar (according
+                # to the corresponding formal argument) but is not a
+                # Reference or a Literal as we don't know whether the result
+                # of any given expression is or is not an array.
```

Comment:
You can add `# pylint: disable=cell-var-from-loop` to remove the pylint warning below

Reply:
Done.